### PR TITLE
[WIP] Add topic coherence pipeline to gensim

### PR DESCRIPTION
**Pull request by dsquareindia**
Addresses #278.
@tmylk @piskvorky is this alright?

### CODE REVIEW

**Comment by piskvorky** [comment]
What is "segmentation"?

Can you add some background info / motivation?

**Comment by dsquareindia** [comment]
@piskvorky sorry I forgot to tag the issue before. I've tagged it now.

**Comment by piskvorky** [comment]
Ah, nice! :-)

@tmylk please review

**Comment by dsquareindia** [comment]
@piskvorky @tmylk I've added the API I was thinking of for adding the topic coherence pipeline. I've currently implemented only the U_mass topic coherence (still have to work a lot on it though) using this pipeline. Sorry for not adding the tests yet. Just wanted to show you what kind of an API I was thinking of. Is this API fine?

**Comment by piskvorky** [comment]
Looks ok to me conceptually (but have a look at `LdaModel.get_topic_terms()` method).

File: gensim/confirmation_measure.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+"""
+
+import logging
+import numpy as np
+
+logger = logging.getLogger(__name__)
+
+EPSILON = 1e-12 # Should be small. Value as suggested in paper.
+
+def Log_Conditional_Probability(segmented_topics, per_topic_probability):
```
Comment (piskvorky) [inline gensim/confirmation_measure.py]: PEP8: function names are lowercase.

File: gensim/models/coherencemodel.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    coherence.
+
+    The main methods are:
+
+    1. constructor, which initializes the four stage pipeline by accepting a coherence measure,
+    2. the ``get_coherence()`` method, which returns the topic coherence.
+
+    Model persistency is achieved via its load/save methods.
+    """
+    def __init__(self, corpus, topics, coherence='u_mass'): # Have to validate coherence input. Check for invalid pipeline methods.
```
Comment (piskvorky) [inline gensim/models/coherencemodel.py]: Always put `FIXME` around notes-to-self. That makes it easy to grep them & make sure we don't forget about it before merge.

**Comment by dsquareindia** [comment]
I've added initial tests for the segmentation module. Although I'm still thinking about what kind of inputs can I get in the segmentation module (or what different outputs can the users use from LDA to input into the pipeline).

File: gensim/segmentation.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+"""
+This module contains functions to perform segmentation on a list of topics.
+"""
+
+import logging
+import numpy as np
+
+logger = logging.getLogger(__name__)
+
+def s_one_pre(topics):
```
Comment (dsquareindia) [inline gensim/segmentation.py]: Add a model argument here. Handle it inside

File: docs/notebooks/u_mass_tutorial.ipynb [hunk 4]
```diff
@@ -0,0 +1,173 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Demonstration for the `u_mass` topic coherence using topic coherence pipeline"
```
Comment (dsquareindia) [inline docs/notebooks/u_mass_tutorial.ipynb]: @tmylk I've added an example notebook for `u_mass` topic coherence. Is this alright?

**Comment by dsquareindia** [comment]
@tmylk @piskvorky I have updated this PR with the latest version. I have also added a `U_mass` tutorial notebook. Could you please have a look at this intermediate version briefly?

**Comment by dsquareindia** [comment]
Added initial draft of `c_v` topic coherence (the best one from the paper).

File: gensim/models/coherencemodel.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        self.topics = self._get_topics()
+        self.coherence = coherence
+        # Set pipeline parameters:
+        if self.coherence == 'u_mass':
+            self.seg = segmentation.s_one_pre
+            self.prob = probability_estimation.p_boolean_document
+            self.conf = direct_confirmation_measure.log_conditional_probability
+            self.aggr = aggregation.arithmetic_mean
+
+        elif self.coherence == 'c_v':
```
Comment (tmylk) [inline gensim/models/coherencemodel.py]: 'c_v' should be a constant to avoid duplication

**Comment by dsquareindia** [comment]
I've added `c_v` coherence to the existing notebook.

**Comment by tmylk** [comment]
Thanks! Why doesn't pyldavis show on github?

**Comment by dsquareindia** [comment]
Yeah even I don't know why... Have to follow the nbviewer link to view it fully :/

**Comment by dsquareindia** [comment]
Added tests for checking mathematical correctness of the `direct_confirmation_measures` module.

File: gensim/models/coherencemodel.py (hunk 5 above)
Comment (dsquareindia) [inline gensim/models/coherencemodel.py]: Sorry can you please elaborate on this a bit more?

**Comment by dsquareindia** [comment]
Added mathematical correctness test for `indirect_confirmation_measure` module.

**Comment by dsquareindia** [comment]
I've modified the ipython notebook a bit. Hopefully it's more human-interpretable now :smile:

**Comment by dsquareindia** [comment]
Added introduction to the notebook. @tmylk can you please review?

**Comment by tmylk** [comment]
@dsquareindia Could you move the new files you created from gensim/*.py to gensim/topic_coherence/*.py ? These modules are only needed for topic coherence and shouldn't be in root.

**Comment by dsquareindia** [comment]
@tmylk done. Could you please check `topic_coherence/__init__.py`? Not quite sure about that.

**Comment by dsquareindia** [comment]
@tmylk tests pass now. Just used `HashDictionary` instead of `Dictionary`.

**Comment by dsquareindia** [comment]
@tmylk I have added support for `LdaVowpalWabbit` wrapper. You can check out an example [here](https://gist.github.com/dsquareindia/1e8574dcb080206edce5dcba90e93d55)

**Comment by guntherzhao** [comment]
I am trying to use Umass measure to pick the best number of topics, but I do not know what Umass exactly means? About the coherence score, is it the bigger, the better, or just the opposite? Below is the output of my test with Umass measure. How many topics should I pick?
image

**Comment by menshikh-iv** [comment]
@guntherzhao simple rule: more - better, here - 7 topics. 

Our coherence is port of https://github.com/dice-group/Palmetto , you can read more about it on [Palmetto wiki page](https://github.com/dice-group/Palmetto/wiki/Coherences) or http://svn.aksw.org/papers/2015/WSDM_Topic_Evaluation/public.pdf

For future - please use [mailing list](https://groups.google.com/forum/#!forum/gensim) for questions (GitHub only for feature requests & bug reports)

**Comment by guntherzhao** [comment]
@menshikh-iv  Thanks for your help!

**Comment by ZeviAltus** [comment]
> I found this answer and seems to be the opposite: https://stackoverflow.com/a/62037698/7654451. For cv is true the more the better, however with u_mass de more negative values are the better. Am I right?

i think that stack overflow answer is wrong.  greater u_mass values are more coherent.

**Comment by dave-jacques** [comment]
Hi, sorry to resurrect this post but my question relates directly to one of the answers.

@menshikh-iv, you said above:

```
simple rule: more - better, here - 7 topics.
```

which I read as "for Umass, the more negative value the better."

Looking at example graph, 5 topics has the most negative Umass coherence score which I therefore take to be best. 

However your answer states 7 topics (which just so happens to be the least negative value).

I'm hoping you could clarify this point, especially as there seems to be a lot of confusion on this topic out there in the world (e.g. [here](https://stackoverflow.com/questions/62032372/coherence-score-u-mass-18-is-good-or-bad))

Thanks!

**Pull request merged**